ChangeSet
  1.1543 04/03/31 15:28:51 dgibson@wocket.ozlabs.ibm.com +1 -0
  Fix bug in iseries_veth driver which could lead to freeing an skb twice
  under high load.

  drivers/net/iseries_veth.c
    1.18 04/03/31 15:28:20 dgibson@wocket.ozlabs.ibm.com +3 -2
    Fix bug in iseries_veth driver which could lead to freeing an skb twice
    under high load.

diff -Nru a/drivers/net/iseries_veth.c b/drivers/net/iseries_veth.c
--- a/drivers/net/iseries_veth.c	Wed Mar 31 09:55:42 2004
+++ b/drivers/net/iseries_veth.c	Wed Mar 31 09:55:42 2004
@@ -999,7 +999,9 @@
 
 	lpmask = veth_transmit_to_many(skb, lpmask, dev);
 
-	if (lpmask) {
+	if (! lpmask) {
+		dev_kfree_skb(skb);
+	} else {
 		spin_lock_irqsave(&port->pending_gate, flags);
 		if (port->pending_skb) {
 			veth_error("%s: Tx while skb was pending!\n", dev->name);
@@ -1014,7 +1016,6 @@
 		spin_unlock_irqrestore(&port->pending_gate, flags);
 	}
 
-	dev_kfree_skb(skb);
 	return 0;
 }
 
.........................................................................
# vim: syntax=diff

